### PR TITLE
Block not visible products from add to cart

### PR DESCRIPTION
**Pull request by pmclain**
### Description (*)
Return not found exception when adding non visible product to cart.

Depends on #699, because the current configurable product implementation adds the child simple product to the cart instead of the configurable.

### Fixed Issues (if relevant)
1. magento/graphql-ce#733

### Manual testing scenarios (*)
1. Create product with visibility set to "Not Visible Individually"
2. Create empty cart
3. Attempt adding sku created in step one to cart

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### CODE REVIEW

File: app/code/Magento/QuoteGraphQl/Model/Cart/AddSimpleProductToCart.php
Comment (naydav) [inline app/code/Magento/QuoteGraphQl/Model/Cart/AddSimpleProductToCart.php]: Hi @pmclain,
Looks like we need use model like it was in
https://github.com/magento/graphql-ce/pull/752/files#diff-dbe98b6f5d79c40dbf020c0865de14eeR53

Thanks

**Review by naydav — CHANGES_REQUESTED** [review]

File: app/code/Magento/QuoteGraphQl/Model/Cart/AddSimpleProductToCart.php
Comment (pmclain) [reply app/code/Magento/QuoteGraphQl/Model/Cart/AddSimpleProductToCart.php]: @naydav Good catch. I noticed the visibility in the linked PR would return an error is sending a product with search only visibility. The update to this PR uses [getVisibleInSiteIds](https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Catalog/Model/Product/Visibility.php#L80-L83) instead and I've opened https://github.com/magento/graphql-ce/pull/760 making a similar adjustment to send friend.

**Review by naydav — APPROVED** [review]

**Comment by naydav** [comment]
Hi @pmclain,
Could you check failed API-functional tests?
Thanks

**Comment by magento-engcom-team** [comment]
Hi @naydav, thank you for the review.
[ENGCOM-5417](https://jira.corp.magento.com/browse/ENGCOM-5417) has been created to process this Pull Request
:eight_spoked_asterisk: @naydav, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by pmclain** [comment]
@naydav This PR depends on https://github.com/magento/graphql-ce/pull/699 and the functional tests will not pass until the configurable product implementation no longer adds not-visible simple products directly to carts.

**Comment by pmclain** [comment]
@naydav I merged upstream after #699 was merged. This resolved the integration test failures.

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=756) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/graphql-ce?pullRequest=756) before we can accept your contribution.<br/>**1** out of **8** committers have signed the CLA.<br/><br/>:white_check_mark: pmclain<br/>:x: SeruyV<br/>:x: Bohdan Shevchenko<br/>:x: victorpetryk<br/>:x: ameysar<br/>:x: zakdma<br/>:x: OlgaVasyltsun<br/>:x: DianaRusin<hr/>**Bohdan Shevchenko, OlgaVasyltsun** seem not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

**Comment by TomashKhamlai** [comment]
After merging the latest mainline I see that schema was slightly changed and now I have some troubles when I add a configurable product to the cart. It still could be not caused by your code and could be hard to reproduce. I will strike out all the steps that are not important from my point of view.

Preconditions:
1. ~2 root categories~
image

2. ~1 website, 2 stores, 2 store views~

Web Site | Store | Store View
-- | -- | --
Main Website</br>(Code: base) | Additional Store</br>(Code: additional)</br>(Root Category: Root_Category) | Test Store View</br>(Code: test)
Main Website</br>(Code: base) | Main Website Store</br>(Code: main_website_store)</br>(Root Category: Default_Category) | Default Store View</br>(Code: default)
3. Create simple product with SKU **simple-product-2-not-visible-black**.
4. Create configurable product with SKU **configurable-product-1**.
Weight 0.1, Price 0
Configurable options based on colors:
black, white, red
qty: 100500, prices: 2.44, 2.66, 8.00
5. When saving choose to create a new attribute set based on the current (Default) attribute set
6. Using **Actions** column disable the white configurable options and choose manually the option for black color. Save the product.
7. Generate empty cart for the Guest
8. Add product to the Cart
```
mutation {
  addConfigurableProductsToCart(
    input: {
      cart_id: "BK0H5yEs0KyfZED7wrEk0aIDHNGaLOG0"
      cart_items: {
        data: {
          quantity: 2
          sku: "configurable-product-1"
        }
        configurable_attributes: {
          id: 93
          value: 4
        }
      }
    }
  ) {
    cart {
      items {
        id
        product {
          color
          sku
        }
        quantity
      }
    }
  }
}
```
Actual result:
```
{
  "errors": [
    {
      "debugMessage": "SQLSTATE[23000]: Integrity constraint violation: 1452 Cannot add or update a child row: a foreign key constraint fails (`graphqlce`.`quote_item_option`, CONSTRAINT `QUOTE_ITEM_OPTION_ITEM_ID_QUOTE_ITEM_ITEM_ID` FOREIGN KEY (`item_id`) REFERENCES `quote_item` (`item_id`) ON DELETE CASCADE), query was: INSERT INTO `quote_item_option` (`product_id`, `code`, `value`) VALUES (?, ?, ?)",
      "message": "Internal server error",
      "category": "internal",
      "locations": [
        {
          "line": 2,
          "column": 3
        }
      ],
      "path": [
        "addConfigurableProductsToCart"
      ]
    }
  ],
  "data": {
    "addConfigurableProductsToCart": null
  }
}
```
Can be useful:
image
</br>
image
</br>
image

**Comment by TomashKhamlai** [comment]
Could be related to #805

**Comment by TomashKhamlai** [comment]
I can confirm that it is related. I was confused with the error message.

**Review by lenaorobei — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @lenaorobei, thank you for the review.
[ENGCOM-5417](https://jira.corp.magento.com/browse/ENGCOM-5417) has been created to process this Pull Request

**Comment by lenaorobei** [comment]
@pmclain While testing this PR we discovered that this validation caused issue with adding `not visible individually` and `not visible` products to cart, although the scenario is different for simple and simple as part of configurable. It is related to https://github.com/magento/graphql-ce/issues/947
Will appreciate If you could compare how product visibility is validated in the checkout flow using REST.

@nrkapoor could you please confirm expected behaviour in the issue https://github.com/magento/graphql-ce/issues/733. Please not, that product is not visible does not mean that it is disabled.

**Comment by lenaorobei** [comment]
Closing this PR since it the fix requires additional refactoring mentioned in https://github.com/magento/graphql-ce/issues/947

**Pull request closed without merging**
